raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -188,6 +189,22 @@ public void resetBeginQuorumEpochTimer(long currentTimeMs) {
         beginQuorumEpochTimer.reset(beginQuorumEpochTimeoutMs);
     }
 
+    public void updateLastReceivedFetchRequest(ReplicaKey replicaKey, long currentTimeMs) {
+        beginQuorumEpochTimer.update(currentTimeMs);
+        lastFetchRequestMs.put(replicaKey, beginQuorumEpochTimer.currentTimeMs());
+    }
+
+    public Set<ReplicaKey> needSendBeginQuorumRequestNodes(long currentTimeMs) {
+        Set<ReplicaKey> replicaKeys = new HashSet<>();
+        beginQuorumEpochTimer.update(currentTimeMs);
+        for (Map.Entry<ReplicaKey, Long> entry : lastFetchRequestMs.entrySet()) {
+            if (beginQuorumEpochTimer.currentTimeMs() - entry.getValue() >= beginQuorumEpochTimeoutMs) {
```

Comment:
From the code above at KafkaRaftClient#L3059,

```java
.filter(key -> !needToSendBeginQuorumRequest.contains(key))
```

 I think the condition should be changed to:
```suggestion
            if (beginQuorumEpochTimer.currentTimeMs() - entry.getValue() < beginQuorumEpochTimeoutMs) {
```

and we should rename the method to `skipSendBeginQuorumRequestNodes`.

Reply:
I also use `needToSendBeginQuorumRequest` and change filter condition because Positive logic is easier to understand.